### ISSUE TITLE
Change Index buffer with another Index buffer

### ISSUE THREAD

**Issue by ToneTC** [bo3b/3Dmigoto#139]
Recently discovered 3Dmigoto and am hoping to get some noob friendly advice. So I have a IB I want to exchange with another IB, and both of them use the same vs and ps. Can this operation be done and if so, can a example script be written up?

After watching DarkStarSword's youtube video on costume mods I feel like I have only dipped my toe into using this tool. If there are any additional "how-to" resources on using #Dmigoto, I would greatly appreciate it!

**Comment by DarkStarSword** [comment]
Replacing index buffers is covered in the costume modding tutorial, starting around this time index:
https://youtu.be/zWE0xP4MgR8?t=1853

    [ShaderRegex]
    ; ShaderRegex with no pattern matches every shader (other than those in
    ; ShaderFixes) of the below versions. Would be better for performance to use
    ; ShaderOverride and specify specific shader hashes, but there can be
    ; multiple shaders (e.g. shadow pass) and not always easy to find them all
    shader_model = vs_4_0 vs_5_0
    ; Match objects based on vertex buffer hash (usually better choice than ib):
    checktextureoverride = vb0
    
    [TextureOverrideBar]
    hash = <hash of vertex buffer to match>
    ; Replace currently bound index buffer with our custom one:
    ib = ResourceReplacementIB
    ; Flag the original draw call to be skipped, and inject our own:
    handling = skip
    drawindexed = auto
    
    [ResourceReplacementIB]
    ; Load custom index buffer from binary file:
    filename = replacement.ib
    type = buffer
    format = DXGI_FORMAT_R16_UINT

Note that while I've omitted it above, it's often a good idea to back up the original index buffer before replacing it, and restoring it afterwards, in case it's needed by a following draw call (e.g. this is needed in DOAXVV to prevent shadows breaking, though in that case I do the backup and restore in the ShaderOverrides so that each individual mod doesn't have to remember).

    [TextureOverrideBar]
    hash = <hash of vertex buffer to match>
    ;
    ; Back up original Index buffer (ref/copy heuristic would pick copy in this case
    ; to snapshot the buffer's current contents - override to ref for performance):
    ResourceBackupIB = ref ib
    ;
    ; As before...
    ib = ResourceReplacementIB
    handling = skip
    drawindexed = auto
    ;
    ; Restore original IB (ref/copy heuristic already picks ref in this case, so don't
    ; technically need to override... but doesn't hurt to be explicit in cases where
    ; you know which should be used, and these kind of backup + restore scenarios
    ; should always be done by reference for performance)
    ib = ref ResourceBackupIB

**Comment by DarkStarSword** [comment]
> If there are any additional "how-to" resources on using #Dmigoto, I would greatly appreciate it!
A lot of the documentation was spread around the old 3D Vision geforce forums, particularly in the new release announcement posts for any new features - but nvidia migrated to a new forum system a while back making these near impossible to find (the links on our github releases page are supposed to go right to the relevant posts, but they now all go to the start of the thread) and losing all our screenshots. I've been meaning to hunt all these posts down to move their text back to github, but haven't got around to that yet.

We do have a wiki with some info on some subjects - please don't be afraid to expand this wiki as you learn any tips and tricks (always remember that if you have a question or get stuck on something it stands to reason others will as well, and just after you figure out how to solve it can be the best time to share your knowledge with others):
https://github.com/bo3b/3Dmigoto/wiki

There's a decent collection of bookmarks on 3D Vision modding and 3DMigoto here, but like the release notes above many of these links were effectively broken when nvidia migrated forums:
http://helixmod.blogspot.com/2017/04/collection-of-bookmarked-posts-related.html

**Comment by ToneTC** [comment]
Is there a way around the following error when trying to import the vb and ib .txt's into blender "Only draw calls using a single vertex buffer and a single index buffer are supported for now"

**Comment by DarkStarSword** [comment]
That will need a (long overdue) update to my blender script to handle. I do have a work in progress sitting in a topic branch, but it isn't ready as yet.

**Comment by ToneTC** [comment]
Since I cannot create a replacement ib/vb file using the beldner addon, can I use the data method you mentioned using the binary files? And if so, how?

**Comment by DarkStarSword** [comment]
In principle yes - 3DMigoto doesn't specifically depend on the Blender addon and the buffers could be crafted/edited by other means, though in practice beyond the Blender addon and 3DMigoto's own frame analysis I don't know of any tools that would generate the buffers in the right format (except maybe the game/engine authoring tools if they are available - but even then they won't just hand you the buffers you need). If you are doing something fairly simple you might be able to hex edit the buffer by hand, but you probably wouldn't want to be doing much more than deleting a few triangles for that.

Hmmm, I've just had a thought... it's worth checking the vertex buffer format to see what the extra vertex buffers are used for - if the per-vertex data is spread over multiple files then I'll really need to update my script to handle it, but if the per-vertex data is all located within a single vertex buffer (i.e. the additional vertex buffer(s) are used for per-instance data) you might be able to get the addon to open it by just deleting the extra vertex buffer(s). To check this, open the .vb.txt file in a text editor and look at the info near the top. If every element that lists `InputSlotClass: per-vertex` also has `InputSlot: 0` (or if not 0 at least they need to all match each other) this might work - delete the `*vb1*`, etc files (every number other than the per-vertex InputSlot), then see if the addon will open it. I haven't tried this, but in theory it should work provided the per-vertex data is all in the same file.


BTW I didn't ask before, but why are you editing the index buffer specifically and not the vertex buffer? In case you don't know the difference between the two - it's analogous to those "join the dots" books, where the vertex buffer are the dots on the page and the index buffer are the instructions on which dots should be joined to which other dots to form an image. I've been kind of assuming you are doing something simple like deleting part of a mesh, which can indeed be done by just editing the index buffer (or alternatively altering the draw calls to specify the first/last indices), but it might help to know a little more about what you are doing to provide better advice. Also, feel free to attach the vb and ib .txt files - I might be able to help and it will give me something to test my blender addon with.


> can I use the data method you mentioned using the binary files

Not sure if it's what you meant, but to be clear, the data= and filename= options are mutually exclusive - either can be used to set the initial data in a buffer, but they can't both be used at the same time. Generally:

filename= is intended to load potentially large or complex buffers (or textures) saved from frame analysis or other applications, such as my blender addon, though in principle it could come from other tools or even hand crafted with a hex editor.

data= is intended for very small, simple cases where a human could hand craft the geometry, e.g. full screen effects and icons are typically drawn with just two triangles where it is fairly easy to hand craft an index buffer specifying which 6 vertices should be connected to form two [counter-]clockwise triangles (e.g. if your vertex buffer contains vertices at the top left, top right, bottom left and bottom right corners of the screen in that order, an index buffer might be constructed with `data=0 1 2 2 1 3`).

The Team Sonic Racing example is as complex as I would ever go using data=, and actually if you have a look at the d3dx.ini for that mod I've got a huge comment above that laying out the index buffer in groups of three (three points / triangle) and explaining what I'm doing (specifically I altered the topology of an existing effect from the game to delete vertices in the dead centre of the screen that were causing problems, and merging the middle four triangles that were connected to these vertices into two wider triangles that skipped over them).

**Comment by ToneTC** [comment]
The game is MortalKombat11. What I am currently trying to do is see if I can swap things like hair, which have unique index buffer hashes, but all hair share the same vertex buffer hash. I was able to work out how to hide elements with ease via your video.

Here is a archive of a frame analysis for one of the outfits with a different hair [archive suffix of '-mk3']. The ib hash is 10724709, the vb hash is 7d2d92607f6c7bdf. This second archive [archive suffix of '-mkc'] has ib hash b0039540 and with the same exact vb hash of 7d2d92607f6c7bdf.

[FrameAnalysis-2020-12-22-102219-mk3.zip](https://github.com/bo3b/3Dmigoto/files/5730696/FrameAnalysis-2020-12-22-102219-mk3.zip)
[FrameAnalysis-2020-12-22-102633-mkc.zip](https://github.com/bo3b/3Dmigoto/files/5730697/FrameAnalysis-2020-12-22-102633-mkc.zip)

**Comment by DarkStarSword** [comment]
7d2d92607f6c7bdf is the hash of the pixel shader, not vertex buffer (also, be aware that vb=vertex buffer and vs=vertex shader are not the same thing). Odd that there doesn't seem to be a corresponding vb file for these meshes in the dump (so it's not actually that there are too many vertex buffers for the blender addon to handle as I had assumed - it's that there are too few), because DrawIndexed surely cannot be used without at least one...

Looking at the log file for one of these I see:

    000101 VSSetShader(pVertexShader:0x00000000CEF3F450, ppClassInstances:0x0000000000000000, NumClassInstances:0) hash=ea57d601e70eadbc
    000101 PSSetShader(pPixelShader:0x00000000C1AA68D0, ppClassInstances:0x0000000000000000, NumClassInstances:0) hash=2c535dd6453ae55b
    ...
    000101 IASetVertexBuffers(StartSlot:0, NumBuffers:32, ppVertexBuffers:0x000000014361C240, pStrides:0x000000014361C340, pOffsets:0x000000014361C340)
    ...
    000101 IASetIndexBuffer(pIndexBuffer:0x00000041FD1D1538, Format:57, Offset:0) hash=10724709
    ...
    000101 DrawIndexed(IndexCount:141174, StartIndexLocation:0, BaseVertexLocation:0)
    000101 3DMigoto Dumping Buffer G:\SteamLibrary\steamapps\common\Mortal Kombat 11\Binaries\Retail\FrameAnalysis-2020-12-22-102219\000101-ib=10724709-vs=ea57d601e70eadbc-ps=2c535dd6453ae55b.buf -> G:\SteamLibrary\steamapps\common\Mortal Kombat 11\Binaries\Retail\FrameAnalysisDeduped\0fa0b226.buf
    000101 3DMigoto Dumping Buffer G:\SteamLibrary\steamapps\common\Mortal Kombat 11\Binaries\Retail\FrameAnalysis-2020-12-22-102219\000101-ib=10724709-vs=ea57d601e70eadbc-ps=2c535dd6453ae55b.txt -> (whoops I have a bug)

Well, that's odd - the IASetVertexBuffers call we see there clears out any vertex buffers left over from previous draw calls (which is fairly common), but there is no additional IASetVertexBuffers call binding any new vertex buffers for this draw call...

I'm at a bit of a loss for how it would be drawing the hair at all given what I can see here... It's possible that it might be getting the vertex data from elsewhere (UE4 is known to do this, passing some of the vertex data in structured buffers instead of vertex buffers - I've had to fix bugs in that engine related to this), so perhaps one of the other resources bound to the vertex shader has the data we need? From the log file there are 5 resources bound as shader resources to the vertex shader and 2 bound as constant buffers that are possible candidates:

    000101 VSSetShaderResources(StartSlot:0, NumViews:1, ppShaderResourceViews:0x000000013FD9F708)
           0: view=0x00000041BE2C1EC0 resource=0x00000041B422B238 hash=9f4b60ba
    000101 VSSetShaderResources(StartSlot:1, NumViews:1, ppShaderResourceViews:0x000000013FD9F708)
           1: view=0x00000041BE2C1680 resource=0x00000041B422C738 hash=5509d86b
    000101 VSSetShaderResources(StartSlot:3, NumViews:1, ppShaderResourceViews:0x000000013FD9F708)
           3: view=0x00000041B3AD7640 resource=0x00000041FD1CDB78 hash=d7c891ec
    000101 VSSetShaderResources(StartSlot:4, NumViews:1, ppShaderResourceViews:0x000000013FD9F6F8)
           4: view=0x00000041BDE0E740 resource=0x00000041BDDC0F38 hash=edc5191e
    000101 VSSetShaderResources(StartSlot:5, NumViews:1, ppShaderResourceViews:0x000000013FD9F6F8)
           5: view=0x00000041B3ACF280 resource=0x00000041FD1D17B8 hash=9f4b60ba
    000101 Map(pResource:0x000000003E898778, Subresource:0, MapType:4, MapFlags:0, pMappedResource:0x000000013FD9F600) hash=0c8934aa
    000101 Unmap(pResource:0x000000003E898778, Subresource:0) hash=0c8934aa
    000101 VSSetConstantBuffers(StartSlot:0, NumBuffers:3, ppConstantBuffers:0x000000013FD9F640)
           0: resource=0x000000003E898778 hash=0c8934aa
           1: resource=0x00000041BDDA3BB8 hash=16d01f37

Cross referencing against ShaderUsage.txt those shader resources are:

    0: <Register orig_hash=9f4b60ba type=Buffer byte_width=379320 usage="DEFAULT" bind_flags="shader_resource unordered_access" cpu_access_flags=0 misc_flags="buffer_structured" stride=12></Register>
    1: <Register orig_hash=5509d86b type=Buffer byte_width=252880 usage="DEFAULT" bind_flags="shader_resource unordered_access" cpu_access_flags=0 misc_flags="buffer_structured" stride=8></Register>
    3: <Register orig_hash=d7c891ec type=Buffer byte_width=505760 usage="DEFAULT" bind_flags="shader_resource unordered_access" cpu_access_flags=0 misc_flags="buffer_structured" stride=4></Register>
    4: <Register orig_hash=edc5191e type=Buffer byte_width=5600 usage="DYNAMIC" bind_flags="shader_resource" cpu_access_flags="write" misc_flags="buffer_structured" stride=112></Register>
    5: <Register orig_hash=9f4b60ba type=Buffer byte_width=379320 usage="DEFAULT" bind_flags="shader_resource unordered_access" cpu_access_flags=0 misc_flags="buffer_structured" stride=12></Register>

Those are all structured buffers and are almost all potential candidates to contain vertex data. We can rule out resource 4 because it's size divided by stride indicates it has 50 entries, far fewer than the ~30,000 or so I see listed in the index buffer. We can rule out the two constant buffers for much the same reason (constant buffers are limited to 4096 entries), plus they are horribly inefficient to dynamically index as would be required to access vertex data. Resources 0, 1 and 5 all contain 31,610 entries, which is very promising as they are large enough to contain all the vertices with a little (but not excessive) room to spare. Resource 3 contains 126,440 entries, which at first seems excessive, but may still be related because divided by 4 again is also 31,610 - so, most likely it is related and just needs a 4x multiplier when indexing it.

All four remaining candidates are marked for unordered access, meaning that other shaders on the GPU can write to them (and indeed earlier in the frame we see compute shader 4e465e62062dc9b0 possibly writing to resources 0, 1 and 5 in dispatch calls 000054 and 000055 - at a guess I wouldn't be surprised if these are involved in a hair physics simulation). Number 3 is marked potentially writable by the GPU, but I do not see it used as such during this particular frame (also, resource 4 that we already ruled out is marked writable by the CPU, though is not updated during this frame).


To get a more complete picture as to what is going on you will probably want to dump out these buffers and the shaders that use them. Add `dump_cb dump_tex` to your analyse_options and run frame analysis again (if it is too slow with these enabled globally you may consider moving the analyse_options into a ShaderOverride to only dump resources used with this vertex shader, though you will probably want the ones passed to the compute shader as well).

Then hunt down vertex shader ea57d601e70eadbc and compute shader 4e465e62062dc9b0 and attach them here - you can either hunt and dump them with `marking_actions = hlsl asm` and the numpad keys (4/5/6 for vertex shaders, hold period + 1/2/3 for compute shaders). If dumping them, I suggest deleting them from ShaderFixes afterwards, as they may cause rendering or other issues, particularly if dumped as HLSL. Alternatlively you can enable export_hlsl=2 and/or export_shaders=1 before starting the game and find them in ShaderCache.


So, as you may have been able to guess by now, this game is doing something weird/unusual and this could well end up being quite tricky to do what you want (I can't handle this in my blender addon in a generic way - it would need a game specific solution that understands where the game has stored its vertex data).

Or, you might be lucky and just swapping the index buffers around might work - I'm not holding my breath, since that would require that all hair models are sharing the same vertex data, which seems unlikely. It is relatively easy to try though, so worth experimenting with to be sure. The binary .buf versions dumped in frame analysis can be used directly in a [Resource] section's filename=

But my guess is that you will very likley need to swap the vertex data around as well (and even then there is no guarantee that this will work if e.g. their vertex groups don't match up), so you will need these structured buffers - only, it might not even be enough to swap them when the hair is being drawn since the GPU appears to write some of them earlier (or it might be - can't say for sure without more information) - you might end up needing to swap them out when the compute shader runs.

**Comment by ToneTC** [comment]
First, thank you so very much for the detailed answers and sorry about providing the pixel shader hash. So I hunted for the two hashes and are providing them in the archive below. I have also made new frame analysis with your suggested settings, but now they're +8gb each, so I will provide a text file of all the files generated by the dump instead of the folder itself. Doing the analysis globally only takes like 30 seconds for the operation to finish- but you said I need to find the vertex data and I have no idea what that would look like.

The archive is also going to include my botched attempt to understand how to swap the index buffers. I really wish this was as simple as hunting other assets and setting them to skip lol.
[Hunted.zip](https://github.com/bo3b/3Dmigoto/files/5733876/Hunted.zip)

**Comment by DarkStarSword** [comment]
TL;DR: I have some bad news - the skeletons aren't compatible between hair meshes, which would seriously complicate swapping them... but I wrote out most of this post before discovering that, so:

Nice - the shaders have their reflection information intact so we see the resource names :) From the vertex shader:

    // StructuredBuffer<VSInputPosition> PositionBuffer : register(t0);
    // StructuredBuffer<VSInputTangent> TangentBuffer : register(t1);
    // StructuredBuffer<uint> TexCoords : register(t3);
    // StructuredBuffer<FDismembermentRenderEllipsoid> Ellipsoids : register(t4);
    // StructuredBuffer<FSkinningPosition> LocalSpacePositions : register(t5);

That pretty much confirms that these structured buffers contain the information that would normally be in a vertex buffer, with (as predicted) the exception of the Ellipsoids buffer in vs-t4 (which I have not looked at in any detail and at this point have no clue what it's for, but I fully expect it come back and bite us later).

We also see the structure layouts (except for TexCoords, which is directly using a primitive uint data type instead of a struct), which are all pretty much as we would expect:

    // struct VSInputPosition
    // {
    //     float3 Position;               // Offset:    0
    // };
    //
    // struct VSInputTangent
    // {
    //     uint TangentX;                 // Offset:    0
    //     uint TangentZ;                 // Offset:    4
    // };
    // struct FSkinningPosition
    // {
    //     float3 Position;               // Offset:    0
    // };

If you end up modifying these buffers yourself in Blender (which you can hopefully avoid if you just intend to swap them) you would need to pay special attention to the TexCoords and VSInputTangent buffers - the fact that these are stored as uints instead of the expected float types is a huge red flag that the game has used some unusual encoding scheme on these. Looking at the code it is clear that the TexCoords (AKA UVs) are packed as two half precision 16bit floats per 32bit uint, and there are multiple texcoords to unpack per vertex (hence the 4x multiplier we noted earlier). The encoding of the TangentBuffer is even more unusual - if I'm reading the code right they have multiplied each axis by 511, cast them to integers, then taken the bottom 10 bits of each and bit shifted them to fit inside a single 32bit value for... reasons I guess :-/

Now, since we already know that at least some of these buffers may have been modified on the GPU we should look at the compute shader, starting with the shader resources and unordered access views that it uses:

    // StructuredBuffer<FDualQuaternion> DualQuatsBuffer : register(t0);
    // StructuredBuffer<float4x3> BoneMatrices : register(t1);
    // StructuredBuffer<FSkinningPosition> InputPositionBuffer : register(t2);
    // StructuredBuffer<FSkinningTangent> InputTangentBuffer : register(t3);
    // StructuredBuffer<FSkinningWeightsAndIndices> InputWeightBuffer : register(t4);
    // StructuredBuffer<FSkinningBlend> InputDQBlendWeightBuffer : register(t5);
    // RWStructuredBuffer<FSkinningPosition> OutputPositionBuffer : register(u0);
    // RWStructuredBuffer<FSkinningTangent> OutputTangentBuffer : register(u1);

Ok, based on these inputs + outputs it looks like this compute shader is applying a combination of bone matrices and quaternion rotations to the hair to calculate the posed position of each vertex (I don't see anything to suggest that there is any sort of physics sim happening in this shader that I thought might be the case earlier). This skinning is normally done in a vertex shader, but I guess they needed to render the hair in multiple passes during the frame and opted to do the skinning once in the compute shader rather than every time the vertex shader is involked (why do no devs ever use the stream output facility for this?).

Some of these inputs we saw in the vertex shader:

InputPositionBuffer in cs-t2  -->  LocalSpacePositions in vs-t5. This is not written to by the compute shader (which I should have spotted earlier if I'd looked at the log a little closer), so this is our neutral pose vertex position input. We will need to replace this one.

OutputPositionBuffer in cs-u0  -->  PositionBuffer in vs-t0. This is written (and not read) from the compute shader, so will contain the skinned vertex positions after the compute shader has been run. While we do not need to replace this from disk, we may possibly still need to swap it out for another buffer if it is too small for the replacement hair.

OutputTangentBuffer in cs-u1  -->  TangentBuffer in vs-t1. This is also written (and not read) from the compute shader, and will contain the skinned tangents (possibly used for e.g. lighting / reflection calculations) after the compute shader has been run. As above, we don't need to replace this from disk, but may need to swap it out if it is too small.

In addition, the compute shader has InputWeightBuffer and InputDQBlendWeightBuffer, which both will contain per-vertex data (the former will contain the vertex groups & weights, the latter has the equivelent for their quaternion rotations). We will need to swap these, but - **and this is critical - since these buffers relate the individual vertices to their place on the hair skeleton, we really need to hope and prey that the hair skeletons are compatible between hair meshes, and that the vertex/quaternion groups are in the same order between meshes.** Unfortunately, this does not appear to be the case:

The remaining inputs DualQuatsBuffer and BoneMatrices contain the transformations for each bone (or at least the latter does - I'll focus on it for now), so if the skeletons are compatible I would expect these buffers to be the same size regardless of which hair is being drawn. Unfortunately, I can already see from the log file that this is not the case:

e.g. looking at the BoneMatrices buffer, in the mk3 log + ShaderUsage files for hair 10724709 I see (noting down both compute shader invocations that wrote to the same position buffer - I guess the hair must be being drawn in two pieces, perhaps the short one might be a scalp cap or maybe there's a ponytail or similar):

    000054 CSSetShaderResources(StartSlot:1, NumViews:1, ppShaderResourceViews:0x000000013FD9FB10)
           1: view=0x00000041ACB12AC0 resource=0x00000041BDD8AF78 hash=c63c764c
           = <Register orig_hash=c63c764c type=Buffer byte_width=336 usage="DYNAMIC" bind_flags="shader_resource" cpu_access_flags="write" misc_flags="buffer_structured" stride=48></Register>
    000054 Dispatch(ThreadGroupCountX:479, ThreadGroupCountY:1, ThreadGroupCountZ:1)
    ...
    000055 CSSetShaderResources(StartSlot:1, NumViews:1, ppShaderResourceViews:0x000000013FD9FB10)
           1: view=0x00000041BE2C9BC0 resource=0x00000041B422D6F8 hash=be5d0bc8
           = <Register orig_hash=be5d0bc8 type=Buffer byte_width=48 usage="DYNAMIC" bind_flags="shader_resource" cpu_access_flags="write" misc_flags="buffer_structured" stride=48></Register>
    000055 Dispatch(ThreadGroupCountX:16, ThreadGroupCountY:1, ThreadGroupCountZ:1)

While looking for the equivelent buffers in the mkc log file for hair b0039540 I see:

    000052 CSSetShaderResources(StartSlot:1, NumViews:1, ppShaderResourceViews:0x000000013FD9FB10)
           1: view=0x00000041BDE04F40 resource=0x00000041BDDC5638 hash=9f8f1ad5
           =<Register orig_hash=9f8f1ad5 type=Buffer byte_width=192 usage="DYNAMIC" bind_flags="shader_resource" cpu_access_flags="write" misc_flags="buffer_structured" stride=48></Register>
    000052 Dispatch(ThreadGroupCountX:101, ThreadGroupCountY:1, ThreadGroupCountZ:1)
    ...
    000053 CSSetShaderResources(StartSlot:1, NumViews:1, ppShaderResourceViews:0x000000013FD9FB10)
           1: view=0x00000041B39F0140 resource=0x00000041BDD9C2F8 hash=3bdd0865
           = <Register orig_hash=3bdd0865 type=Buffer byte_width=1248 usage="DYNAMIC" bind_flags="shader_resource" cpu_access_flags="write" misc_flags="buffer_structured" stride=48></Register>
    000053 Dispatch(ThreadGroupCountX:628, ThreadGroupCountY:1, ThreadGroupCountZ:1)

With each of those four buffers being very different sizes I have to conclude that each hair must have its own skeleton, which presents a serious challenge for swapping the hair meshes with 3DMigoto.

This is not necessarily a complete dead end and if you really wanted to keep pushing you might be able to find an acceptable solution (e.g. the simplest solution might be to disregard the vertex weights entirely if you don't mind the replaced hair being stiff, or you might find that there are enough bones in common between the skeletons to be able to make do those and allow for the rest to be missing, or theoretically you could make your own more complete skeleton and code up a custom shader to calculate the transformation matrices for it), but it will be an uphill battle and this is probably where I would stop.

You may have better luck developing a modding tool specifically for the game - if you can pull apart their file formats you may be able to swap the meshes along with the skeletons thereby allowing the swapped hair to behave naturally. This is more or less what ended up happening in DOA6 where early mods used 3DMigoto due to the benefit of having a generic tool that already existed and would work right away out of the box, but later on someone developed a tool specific to the game that could swap things that are not visible to DirectX & 3DMigoto like the skeletons (although funnily enough they still use my Blender addon because they were able to pull the same vertex and index buffers from the game files that 3DMigoto would pull from DirectX and they didn't see any reason to reinvent the process of importing / exporting to Blender, plus I wrote a nice soft body fixup script for them that was necessary for any mods that altered geometry within the soft body physics regions of the mesh).
